foxglove_bridge_base/include/foxglove_bridge/server_interface.hpp

```diff
@@ -17,6 +17,63 @@ constexpr size_t DEFAULT_SEND_BUFFER_LIMIT_BYTES = 10000000UL;  // 10 MB
 
 using MapOfSets = std::unordered_map<std::string, std::unordered_set<std::string>>;
 
+template <typename IdType>
+class ExeptionWithId : public std::runtime_error {
+public:
+  explicit ExeptionWithId(IdType id, const std::string& what_arg)
+      : std::runtime_error(what_arg)
+      , _id(id) {}
+
+  IdType id() const {
+    return _id;
+  }
+
+private:
+  IdType _id;
+};
+
+class ChannelError : public ExeptionWithId<ChannelId> {
+  using ExeptionWithId::ExeptionWithId;
+};
+class UnknownChannelError : public ChannelError {
+  using ChannelError::ChannelError;
+};
+class AlreadySubscribedError : public ChannelError {
+  using ChannelError::ChannelError;
+};
+class NotSubscribedError : public ChannelError {
+  using ChannelError::ChannelError;
+};
+
+class ClientChannelError : public ExeptionWithId<ClientChannelId> {
+  using ExeptionWithId::ExeptionWithId;
+};
+class UnsupportedEncodingError : public ClientChannelError {
+  using ClientChannelError::ClientChannelError;
+};
+class NotAdvertisedError : public ClientChannelError {
+  using ClientChannelError::ClientChannelError;
+};
+class AlreadyAdvertisedError : public ClientChannelError {
+  using ClientChannelError::ClientChannelError;
+};
+class ClientChannelSchemaError : public ClientChannelError {
+  using ClientChannelError::ClientChannelError;
+};
+
+class ServiceError : public ExeptionWithId<ServiceId> {
+  using ExeptionWithId::ExeptionWithId;
+};
+class ServiceUnknownError : public ServiceError {
+  using ServiceError::ServiceError;
+};
+class ServiceUnavailableError : public ServiceError {
+  using ServiceError::ServiceError;
+};
+class ServiceSchemaError : public ServiceError {
+  using ServiceError::ServiceError;
+};
```

Comment:
Since we already have a detailed message as part of the error, I wonder if we really need a separate subclass for each type of error, or could you just use ServiceError for all of these?

Reply:
This is part of the public API, so not sure what error granularity is best here. I guess I will remove all error types that are not explicitly caught in the server code